src/Text/Gigaparsec/Internal/Token/Space.hs

```diff
@@ -0,0 +1,233 @@
+This type is concerned with special treatment of whitespace.
+
+For the vast majority of cases, the functionality within this object shouldn't be needed, 
+as whitespace is consistently handled by lexeme and fully. 
+However, for grammars where whitespace is significant (like indentation-sensitive languages), 
+this object provides some more fine-grained control over how whitespace is consumed by the parsers within lexeme.
+-}
+type Space :: *
+data Space = Space {
+  {-|
+  Skips zero or more (insignificant) whitespace characters as well as comments.
+
+  The implementation of this parser depends on whether 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true: 
+  when it is, this parser may change based on the use of the alter combinator. 
+
+  This parser will always use the hide combinator as to not appear as a valid alternative in an error message: 
+  it's likely always the case whitespace can be added at any given time, but that doesn't make it a useful suggestion unless it is significant.
+  -}
+    whiteSpace :: !(Parsec ())
+  {-|
+  Skips zero or more comments.
+
+  The implementation of this combinator does not vary with 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent'. 
+  It will use the hide combinator as to not appear as a valid alternative in an error message: 
+  adding a comment is often legal, 
+  but not a useful solution for how to make the input syntactically valid.
+  -}
+  , skipComments :: !(Parsec ())
+  {-|
+  This combinator changes how lexemes parse whitespace for the duration of a given parser.
+
+  So long as 'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true, 
+  this combinator will be able to locally change the definition of whitespace during the given parser.
+
+  === __Examples__
+  * In indentation sensitive languages, the indentation sensitivity is often ignored within parentheses or braces. 
+  In these cases, 
+  @parens (alter withNewLine p)@ 
+  would allow unrestricted newlines within parentheses.
+  -}
+  , alter :: forall a. Desc.CharPredicate -> Parsec a -> Parsec a
+  {-|
+  This parser initialises the whitespace used by the lexer when 
+  'Text.Gigaparsec.Token.Descriptions.whiteSpaceIsContextDependent' is true.
+
+  The whitespace is set to the implementation given by the lexical description.
+  This parser must be used, by fully or otherwise, 
+  as the first thing the global parser does or an UnfilledRegisterException will occur.
+
+  See 'alter' for how to change whitespace during a parse.
+  -}
+  , initSpace :: Parsec ()
+  }
+
+---------------------------------------------------------------------------------------------------
+-- Whitespace parsers setup
+
+{-|
+A global ref to the mapping from threads to their corresponding whitespace parsers.
+
+Implemented in terms of `unsafePerformIO` -- use with caution!
+-}
+-- TODO: use a HashMap instead of a Map?
+{-# NOINLINE wsImplMap #-}
+wsImplMap :: IORef (Map ThreadId (Parsec ()))
```

Comment:
This map is memory leaking threads, it would be good to use Weak thread IDs, which wouldn't retain the thread; then when we try and initialise our whitespace we clean away any dead references from the map.

Reply:
Is it safe to put an `Ord` or `Enum`/`hashable` instance on weak threadIds? Or will it need to use an unordered map